app/models/app_remote_kill.rb

```diff
@@ -0,0 +1,77 @@
+class AppRemoteKill < ActiveRecord::Base
+  belongs_to :app
+  belongs_to :split
+
+  attribute :first_bad_version, :app_version
+  attribute :fixed_version, :app_version
+
+  validates :app, :split, :reason, :override_to, :first_bad_version, presence: true
+  validates :reason, uniqueness: { scope: %i(app split) }
+
+  validate :override_to_must_exist
+  validate :fixed_version_must_be_greater_than_first_bad_version
+  validate :must_not_overlap_existing
+
+  scope :affecting, ->(app_build, override: false, overridden_at: nil) do
+    where(
+      arel_table[:app_id].eq(app_build.app_id)
+      .and(arel_table[:first_bad_version].lteq(app_build.version))
+      .and(arel_fixed_version_is_null_or_greater_than(app_build.version))
+      .and(
+        Arel::Nodes::Grouping.new(
+          arel_is_false(override)
+          .or(arel_table[:updated_at].gt(overridden_at))
+        )
+      )
+    )
+  end
+
+  class << self
+    private
+
+    def arel_is_false(value_or_arel)
+      Arel::Nodes::Grouping.new(Arel::Nodes::False.new).eq(value_or_arel)
+    end
+  end
+
+  scope :overlapping, ->(other) do
+    where(app_id: other.app_id, split_id: other.split_id)
+      .where.not(id: other.id)
+      .where(
+        if other.fixed_version.nil?
+          Arel::Nodes::True.new
+        else
+          arel_table[:first_bad_version].lt(other.fixed_version)
+        end
+        .and(arel_fixed_version_is_null_or_greater_than(other.first_bad_version))
+      )
+  end
+
+  def self.arel_fixed_version_is_null_or_greater_than(version)
+    Arel::Nodes::Grouping.new(
+      arel_table[:fixed_version].eq(nil)
+      .or(arel_table[:fixed_version].gt(version))
+    )
+  end
+
+  def override_to_must_exist
+    return unless split
+
+    errors.add(:override_to, "must exist in split's current variants") unless split.has_variant?(override_to)
+  end
+
+  def fixed_version_must_be_greater_than_first_bad_version
+    return if fixed_version.nil?
```

Comment:
same -- `validate :fixed_version_must_be_greater_than_first_bad_version, if: :fixed_version`?

Reply:
As above, you could do it that way, but I see this as a complete unit - a validator that operates only on extant values.

Reply:
Got it. So the structure is basically:

```ruby
def thing_must_behave
  return if something # nil-check

  errors.add(field, "must behave") if condition # actual validation condition
```

And the point is to keep the nil-check separate from the actual validation condition?

Reply:
I'm not opposed to this as a choice, but I do wish the existing validations DSL / errors API made this a little easier to express without early `return`s everywhere.

Reply:
yeah, agreed.